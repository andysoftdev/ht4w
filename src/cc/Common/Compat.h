/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Compatibility Macros for C/C++.
 * This file contains compatibility macros for C/C++. Always include this
 * file before including any other file!
 */

#ifndef HYPERTABLE_COMPAT_H
#define HYPERTABLE_COMPAT_H

/** @addtogroup Common
 *  @{
 */

// The same stuff for C code
#include "compat-c.h"

#include <cstddef> // for std::size_t and std::ptrdiff_t
#include <memory>

#if _MSC_VER <= 1800

// the chrono nightmare starts here
#define _CHRONO_
#include <boost/chrono/chrono.hpp>
#include <thr/xtimec.h>

namespace std {
  namespace chrono {

    typedef ::boost::chrono::steady_clock steady_clock;
    typedef ::boost::chrono::system_clock system_clock;

    typedef ::boost::chrono::hours hours;
    typedef ::boost::chrono::minutes minutes;
    typedef ::boost::chrono::seconds seconds;
    typedef ::boost::chrono::milliseconds milliseconds;
    typedef ::boost::chrono::microseconds microseconds;
    typedef ::boost::chrono::nanoseconds nanoseconds;

    template<typename Clock, typename Duration = typename Clock::duration>
    using time_point = ::boost::chrono::time_point<Clock, Duration>;

    template<typename Rep, typename Period>
    using duration = ::boost::chrono::duration< Rep, Period>;

    template <typename ToDuration, typename Rep, typename Period>
    inline BOOST_CONSTEXPR
    typename ::boost::enable_if<::boost::chrono::detail::is_duration<ToDuration>, ToDuration>::type
    duration_cast(const duration<Rep, Period>& fd) {
      return ::boost::chrono::detail::duration_cast<
        duration<Rep, Period>, ToDuration>()(fd);
    }

    #define CHRONO_OPERATOR \
      ::boost::chrono::operator

  }

  template<typename _Rep, typename _Period> inline
  ::xtime _To_xtime(const chrono::duration<_Rep, _Period>& _Rel_time){
    // convert duration to xtime
    ::xtime _Xt;
    if (_Rel_time <= chrono::duration<_Rep, _Period>::zero()) {
      // negative or zero relative time, return zero
      _Xt.sec = 0;
      _Xt.nsec = 0;
    }
    else { // positive relative time, convert
      chrono::nanoseconds _T0 =
        chrono::system_clock::now().time_since_epoch();
      _T0 += _Rel_time;
      _Xt.sec = chrono::duration_cast<chrono::seconds>(_T0).count();
      _T0 -= chrono::seconds(_Xt.sec);
      _Xt.nsec = (long)_T0.count();
    }
    return (_Xt);
  }

}

#endif

// C++ specific stuff
#ifndef BOOST_SPIRIT_THREADSAFE
#  define BOOST_SPIRIT_THREADSAFE
#endif

#define BOOST_IOSTREAMS_USE_DEPRECATED
#define BOOST_FILESYSTEM_VERSION 3
#define BOOST_FILESYSTEM_DEPRECATED     1

#define HT_UNUSED(x) static_cast<void>(x)

template<typename T, typename... Ts>
std::unique_ptr<T> make_unique(Ts&&... params) {
  return std::unique_ptr<T>(new T(std::forward<Ts>(params)...));
}

#if defined(__APPLE__) || !defined(_GLIBCXX_HAVE_QUICK_EXIT)
namespace std {
  inline void quick_exit(int status) { _exit(status); }
}
#endif

/** @}*/

#endif // HYPERTABLE_COMPAT_H

