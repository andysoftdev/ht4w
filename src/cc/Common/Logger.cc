/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Logging routines and macros.
 * The LogWriter provides facilities to write debug, log, error- and other
 * messages to stdout. The Logging namespaces provides std::ostream-
 * and printf-like macros and convenience functions.
 */

#include <Common/Compat.h>

#include "String.h"
#include "Logger.h"

#include <iostream>
#include <stdio.h>
#include <stdarg.h>
#include <mutex>
#include "FileUtils.h"

namespace Hypertable { namespace Logger {

static String logger_name;
static LogWriter *logger_obj = 0;
static std::mutex mutex;

void initialize(const String &name) {
  logger_name = name;
}

LogWriter *get() {
  if (!logger_obj)
    logger_obj = new LogWriter(logger_name);
  return logger_obj;
}

void LogWriter::log_string(int priority, const char *message) {
  static const char *priority_name[] = {
    "FATAL",
    "ALERT",
    "CRIT",
    "ERROR",
    "WARN",
    "", // NOTICE
    "INFO",
    "DEBUG",
    "NOTSET"
  };

  std::lock_guard<std::mutex> lock(mutex);
  int saved_errno = errno;

  if (m_test_mode) {
    fprintf(m_file, "%s %s : %s\n", priority_name[priority], m_name.c_str(),
            message);
  }
  else {
    time_t t = ::time(0);

#ifdef _WIN32

    struct tm lt;
    localtime_s(&lt, &t);
    char dateTime[64];
    strftime(dateTime, sizeof(dateTime), "%Y-%m-%d %H:%M:%S", &lt);

    String entry = format("%s %6s %s", 
                          dateTime,
                          priority_name[priority],
                          message);
    fprintf(m_file, "%s\n", entry.c_str());

#else

    fprintf(m_file, "%u %s %s : %s\n", (unsigned)t, priority_name[priority],
            m_name.c_str(), message);

#endif

    for ( const LogSink* ls : logSinks )
      ls->log( priority, message, entry );

  }

#ifndef _WIN32

  flush();

#else

  _fflush_nolock(m_file);

#endif

  if (!m_logfile.empty() && m_file != stdout) {
    if (ftell(m_file) > 1024*1024) {
      fclose(m_file);
      for (int i = 1; i < 9999; ++i) {
        String dest = format("%s.%04d", m_logfile.c_str(), i);
        if (!FileUtils::exists(dest)) {
          FileUtils::rename(m_logfile.c_str(), dest.c_str());
          break;
        }
      }
      m_file = fopen(m_logfile.c_str(), "a+tc");
      if (m_file)
        fseek(m_file, 0, SEEK_END);
    }
  }

  errno = saved_errno;
}

void LogWriter::log_varargs(int priority, const char *format, va_list ap) {
  char buffer[1024 * 16];
  vsnprintf(buffer, sizeof(buffer), format, ap);
  log_string(priority, buffer);
}

void LogWriter::debug(const char *format, ...) {
  va_list ap;
  va_start(ap, format);
  log_varargs(Priority::DEBUG, format, ap);
  va_end(ap);
}

void LogWriter::log(int priority, const char *format, ...) {
  va_list ap;
  va_start(ap, format);
  log_varargs(priority, format, ap);
  va_end(ap);
}

void LogWriter::set_file(const char* logfile) {
  if (logfile && *logfile) {
    std::lock_guard<std::mutex> lock(mutex);
    m_logfile = logfile;
    m_file = fopen(logfile, "a+tc");
    if (m_file)
      fseek(m_file, 0, SEEK_END);
  }
}

void LogWriter::set_file(FILE *file) {
  if (file) {
    std::lock_guard<std::mutex> lock(mutex);
    m_file = file;
  }
}

void LogWriter::add_sink(const LogSink* ls) {
  std::lock_guard<std::mutex> lock(mutex);
  if (ls)
    logSinks.insert(ls);
}

void LogWriter::remove_sink(const LogSink* ls) {
  std::lock_guard<std::mutex> lock(mutex);
  if (ls)
    logSinks.erase(ls);
}

void LogWriter::set_test_mode(int fd) {
  if (fd != -1) {

#ifndef _WIN32

    m_file = fdopen(fd, "wt");

#else

    m_file = fdopen(fd, "w+tc");

#endif

  }
  m_show_line_numbers = false;
  m_test_mode = true;
}

void LogWriter::flush() {
  std::lock_guard<std::mutex> lock(mutex);

#ifndef _WIN32

  fflush(m_file);

#else

  _fflush_nolock(m_file);

#endif
}

void LogWriter::close() {
  std::lock_guard<std::mutex> lock(mutex);

  if (m_file != stdout) {
    fclose(m_file);
    m_file = stdout;
    m_logfile.clear();
  }
}

}} // namespace Hypertable::
