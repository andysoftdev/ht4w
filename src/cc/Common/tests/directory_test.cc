/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3
 * of the License.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include <Common/Compat.h>
#include <Common/directory.h>
#include <Common/Logger.h>
#include <Common/PageArenaAllocator.h>

#include <cstdlib>
#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <stack>
#include <string>
#include <vector>

using namespace Hypertable;
using namespace std;

ofstream out;

template <typename K, typename V, class C = std::less<K>>
void display_directory(typename directory<K,V,C>::value_type &entry) {
  static vector<K> comps;  
  if (entry.level == 0)
    comps.clear();
  else {
    if (comps.size() >= entry.level)
      comps.resize(entry.level-1);
    comps.push_back(entry.key);
  }
  stringstream ss;
  for (const auto &key : comps)
    ss << "/" << key;
  if (entry.isset_value)
    ss << "=" << entry.value;
  else {
    if (ss.str().length() > 1)
      ss << "/";
  }
  if (entry.level == 0)
    ss << "/";
  out << ss.str() << endl;
}


template <class D>
void load_directory(D &dir) {
  vector<string> path;
  path.push_back("a");
  path.push_back("b");
  path.push_back("c");
  path.push_back("1");
  path.push_back("a");
  dir.insert(path, "foo");
  path.pop_back();
  path.push_back("b");
  dir.insert(path, "bar");
  path.pop_back();
  path.pop_back();
  path.push_back("2");
  path.push_back("a");
  dir.insert(path, "how");
  path.pop_back();
  path.push_back("b");
  dir.insert(path, "now");
  path.pop_back();
  path.pop_back();
  path.push_back("3");
  path.push_back("a");
  dir.insert(path, "brown");
  path.pop_back();
  path.push_back("b");
  dir.insert(path, "cow");
}


void test_default_constructor() {
  out << "[default_constructor]" << endl;
  directory<string, string> dir;
  for_each(dir.begin(), dir.end(), display_directory<string, string>);
  HT_ASSERT(dir.empty());
}

void test_copy_constructor() {
  out << "[copy_constructor]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  directory<string, string> dir2(dir);
  for_each(dir2.begin(), dir2.end(), display_directory<string, string>);
  out << "dir is " << (dir.empty() ? "empty" : "not empty") << endl;
  out << "dir2 is " << (dir2.empty() ? "empty" : "not empty") << endl;
  HT_ASSERT(dir.size() == dir2.size());
  HT_ASSERT(dir.max_size() == dir2.max_size());
}

void test_move_constructor() {
  out << "[move_constructor]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  directory<string, string> dir2(std::move(dir));
  for_each(dir2.begin(), dir2.end(), display_directory<string, string>);
  out << "dir is " << (dir.empty() ? "empty" : "not empty") << endl;
  out << "dir2 is " << (dir2.empty() ? "empty" : "not empty") << endl;
  out << "dir.size() == " << dir.size() << endl;
  HT_ASSERT(dir.max_size() == dir2.max_size());
}

void test_range_constructor() {
  out << "[range_constructor]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  for_each(dir.begin(), dir.end(), display_directory<string, string>);

  directory<string, string> dir2(dir.begin(), dir.end());
  for_each(dir2.begin(), dir2.end(), display_directory<string, string>);

  vector<string> path;
  path.push_back("a");
  path.push_back("b");
  path.push_back("c");
  path.push_back("1");
  auto begin_iter = dir.find(path);
  path.pop_back();
  path.push_back("2");
  auto end_iter = dir.find(path);

  directory<string, string> dir3(begin_iter, end_iter);
  for_each(dir3.begin(), dir3.end(), display_directory<string, string>);

  begin_iter = dir.begin();
  end_iter = dir.begin();
  while (end_iter != dir.end()) {
    ++end_iter;
    directory<string, string> dir4(begin_iter, end_iter);
    for_each(dir4.begin(), dir4.end(), display_directory<string, string>);
  }

}

void test_reverse_iteration() {
  out << "[reverse_iteration]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  for (auto riter = dir.rbegin(); riter != dir.rend(); ++riter)
    out << *riter << endl;
}

void test_reverse_insert() {
  out << "[reverse_insert]" << endl;
  vector<string> path;

  directory<string, string> dir;
  path.push_back("a");
  path.push_back("a");
  path.push_back("a");
  path.push_back("a");
  dir.insert(path, "foo");
  auto iter = dir.find(path);

  path.clear();
  directory<string, string> dir2;
  path.push_back("b");
  path.push_back("b");
  path.push_back("b");
  path.push_back("b");
  dir2.insert(path, "bar");

  dir.insert(iter, dir2.rbegin(), dir2.rend());

  for_each(dir.begin(), dir.end(), display_directory<string, string>);

}

void test_copy_assignment() {
  out << "[copy_assignment]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  auto dir2 = dir;
  for_each(dir2.begin(), dir2.end(), display_directory<string, string>);
}

void test_move_assignment() {
  out << "[move_assignment]" << endl;
  directory<string, string> dir;
  load_directory(dir);
  auto dir2 = std::move(dir);
  for_each(dir2.begin(), dir2.end(), display_directory<string, string>);
}


void test_swap() {
  out << "[swap]" << endl;
  directory<string, string> dir, dir2;
  load_directory(dir);
  size_t before_size = dir.size();
  dir.swap(dir2);
  HT_ASSERT(dir.empty());
  swap(dir, dir2);
  HT_ASSERT(dir2.empty());
  HT_ASSERT(dir.size() == before_size);
}


void test_reverse_comparator() {
  out << "[reverse_comparator]" << endl;
  auto cmp = [](const string &lhs, const string &rhs) -> bool { return rhs.compare(lhs) < 0; };
  directory<string, string, decltype(cmp)> dir(cmp);
  load_directory(dir);
  for_each(dir.begin(), dir.end(), display_directory<string, string, decltype(cmp)>);
}


void test_int_keys() {
  out << "[int_keys]" << endl;
  directory<int, string> dir;

  vector<int> path;
  path.push_back(1);
  path.push_back(2);
  path.push_back(3);
  path.push_back(4);
  path.push_back(5);
  dir.insert(path, "foo");
  path.pop_back();
  path.push_back(6);
  dir.insert(path, "bar");
  path.pop_back();
  path.pop_back();
  path.push_back(16);
  path.push_back(38);
  dir.insert(path, "how");
  path.pop_back();
  path.push_back(44);
  dir.insert(path, "now");
  path.pop_back();
  path.pop_back();
  path.push_back(81);
  path.push_back(27);
  dir.insert(path, "brown");
  path.pop_back();
  path.push_back(29);
  dir.insert(path, "cow");
  for_each(dir.begin(), dir.end(), display_directory<int, string>);
}

void test_equals() {
  out << "[equals]" << endl;
  directory<string, string> dir, dir2;
  load_directory(dir);
  load_directory(dir2);
  HT_ASSERT(dir == dir2);
  vector<string> path;
  path.push_back("a");
  path.push_back("b");
  path.push_back("c");
  path.push_back("4");
  path.push_back("a");
  dir.insert(path, "Hello");
  HT_ASSERT(dir != dir2);
  dir2.insert(path, "World");
  HT_ASSERT(dir != dir2);
}

void test_relational() {
  out << "[relational]" << endl;
  vector<string> path;
  {
    directory<string, string> dir, dir2;
    path.push_back("a");
    dir.insert(path, "foo");
    path.clear();
    path.push_back("b");
    dir2.insert(path, "foo");
    HT_ASSERT(dir < dir2);
  }
  {
    directory<string, string> dir, dir2;
    path.clear();
    path.push_back("a");
    dir.insert(path, "bar");
    path.clear();
    path.push_back("a");
    dir2.insert(path, "foo");
    HT_ASSERT(dir < dir2);
    HT_ASSERT(dir <= dir2);
    HT_ASSERT(dir2 > dir);
    HT_ASSERT(dir2 >= dir);
  }
  {
    directory<string, string> dir, dir2;
    path.clear();
    path.push_back("a");
    dir.insert(path, "");
    path.clear();
    path.push_back("b");
    dir.insert(path, "");
    path.push_back("a");
    path.push_back("b");
    dir2.insert(path, "");
    HT_ASSERT(dir < dir2);
    HT_ASSERT(dir <= dir2);
    HT_ASSERT(dir2 > dir);
    HT_ASSERT(dir2 >= dir);
  }
  {
    directory<string, string> dir, dir2;
    path.clear();
    path.push_back("a");
    dir.insert(path, "foo");
    dir2.insert(path, "foo");
    HT_ASSERT(!(dir < dir2));
    HT_ASSERT(dir >= dir2);
    HT_ASSERT(dir <= dir2);
  }
}


void test_alternate_allocator() {
  out << "[alternate_allocator]" << endl;
  ByteArena arena;
  PageArenaAllocator<directory_entry<string, string>, decltype(arena)> alloc(arena);
  directory<string, string, less<string>, decltype(alloc)> dir(alloc);
  load_directory(dir);
  for_each(dir.begin(), dir.end(), display_directory<string, string>);
}


int main(int argc, char **argv) {
  string golden;

  if (argc == 2) {
    if (strcmp(argv[1], "--golden"))
      golden.append(argv[1]);
  }
  else if (argc == 1) {
    cout << "usage: directory_test <golden-file>" << endl;
    cout << " or" << endl;
    cout << "usage: directory_test --golden" << endl;
    quick_exit(EXIT_SUCCESS);
  }

  out.open("directory_test.output", std::ofstream::out);

  test_default_constructor();
  test_copy_constructor();
  test_move_constructor();
  test_range_constructor();
  test_reverse_iteration();
  test_reverse_insert();
  test_copy_assignment();
  test_move_assignment();
  test_swap();
  test_reverse_comparator();
  test_int_keys();
  test_equals();
  test_relational();
  test_alternate_allocator();

  out.close();

  if (golden.empty()) {
    string cmd("cat directory_test.output");
    if (system(cmd.c_str()))
      quick_exit(EXIT_FAILURE);
  }
  else {
    string cmd("diff directory_test.output ");
    cmd.append(golden);
    if (system(cmd.c_str()))
      quick_exit(EXIT_FAILURE);
  }

  quick_exit(EXIT_SUCCESS);
    
}
