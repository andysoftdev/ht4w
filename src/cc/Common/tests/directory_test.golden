[default_constructor]
[copy_constructor]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
dir is not empty
dir2 is not empty
[move_constructor]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
dir is empty
dir2 is not empty
dir.size() == 0
[range_constructor]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
/
/1/
/1/a=foo
/1/b=bar
/
/
/a/
/
/a/
/a/b/
/
/a/
/a/b/
/a/b/c/
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
[reverse_iteration]
 5 b=cow
 5 a=brown
 4 3
 5 b=now
 5 a=how
 4 2
 5 b=bar
 5 a=foo
 4 1
 3 c
 2 b
 1 a
 0 
[reverse_insert]
/
/
/a/
/a/a/
/a/a/a/
/a/a/a/a=foo
/a/a/a/a/b=bar
/a/a/a/b/
/a/a/b/
/a/b/
[copy_assignment]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
[move_assignment]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
[swap]
[reverse_comparator]
/
/a/
/a/b/
/a/b/c/
/a/b/c/3/
/a/b/c/3/b=cow
/a/b/c/3/a=brown
/a/b/c/2/
/a/b/c/2/b=now
/a/b/c/2/a=how
/a/b/c/1/
/a/b/c/1/b=bar
/a/b/c/1/a=foo
[int_keys]
/
/1/
/1/2/
/1/2/3/
/1/2/3/4/
/1/2/3/4/5=foo
/1/2/3/4/6=bar
/1/2/3/16/
/1/2/3/16/38=how
/1/2/3/16/44=now
/1/2/3/81/
/1/2/3/81/27=brown
/1/2/3/81/29=cow
[equals]
[relational]
[alternate_allocator]
/
/a/
/a/b/
/a/b/c/
/a/b/c/1/
/a/b/c/1/a=foo
/a/b/c/1/b=bar
/a/b/c/2/
/a/b/c/2/a=how
/a/b/c/2/b=now
/a/b/c/3/
/a/b/c/3/a=brown
/a/b/c/3/b=cow
